pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+                #from IPython import embed
+                if np.isfinite(centv) is False: debugger.set_trace() #embed()
+                pcen = int(0.5 + centv)
+                mtfit[sz-k] = 0
+        '''
+        jxp_fix = False
+        if jxp_fix:
+            from desispec.bootcalib import trace_crude_init
+            from desispec.bootcalib import trace_fweight as dbtf
+            from pypit import ginga
+            pcen = arcdet[j]
+            img = msarc[pcen-nspecfit:pcen+nspecfit+1, ordcen[arcdet[j], slitnum]-sz:ordcen[arcdet[j], slitnum]+sz+1]
+            rot_img = np.rot90(img,3) - 980.  # Bias!
+            #
+            xcen = np.array([6.6]) # 8.5, 173
+            ypass = 173
+            # Tune-up first
+            for ii in range(4):
+                xcen, xsig = dbtf(rot_img, xcen, ycen=np.array([ypass]).astype(int), invvar=None, radius=2.)
+            xset, xerr = trace_crude_init(rot_img, xcen, ypass, invvar=None, radius=3., maxshift0=0.5, maxshift=0.15, maxerr=0.2)
+            #xcen, xsig = dbtf(rot_img, np.array([6.5,6.5]), ycen=np.array([173,173]).astype(int), invvar=None, radius=2.)
+            # Convert back
+            y0 = pcen+nspecfit
+            ycrude = y0 - xset[:,0]
+            #debugger.set_trace()
+            #cytfit = ytfit.copy()
+            #cytfit[np.where(ytfit < 0)] = np.median(cytfit)
+            #debugger.xplot(np.arange(img.shape[1]), ycrude, cytfit)
+            #
+            #trcdict['save_yt'] = ytfit.copy()
+            assert ycrude.size == ytfit.size
+            ytfit = ycrude
+            #debugger.set_trace()
+        '''
+
+        if offchip:
+            # Don't use lines that go off the chip (could lead to a bad trace)
+            aduse[j] = False
+            badlines += 1
+            trcdict = pad_dict(trcdict)
+            continue
+        for k in range(sz+1-nsmth, sz+1):
+            xtfit[sz-k] = ordcen[arcdet[j], slitnum]-k
+
+        wmask = np.where(mtfit == 0)
+        ytfit[np.where(mtfit == 1)] = maskval
+
+        # Append the trace information into the dictionary
+        trcdict["xtfit"].append(xtfit.copy())
+        trcdict["ytfit"].append(ytfit.copy())
+        trcdict["wmask"].append(wmask[0].copy())
+    trcdict["aduse"] = aduse
+    trcdict["badlines"] = badlines
+    msgs.info("Completed spectral tilt tracing".format(np.sum(aduse)))
+    return trcdict
+
+
+
+def trace_fweight(fimage, xinit, ltrace=None, rtraceinvvar=None, radius=3.):
+    """ Python port of trace_fweight.pro from IDLUTILS
+
+    Parameters:
+    -----------
+    fimage: 2D ndarray
+      Image for tracing
+    xinit: ndarray
+      Initial guesses for x-trace
+    invvar: ndarray, optional
+      Inverse variance array for the image
+    radius: float, optional
+      Radius for centroiding; default to 3.0
+    """
+
+    # Init
+    nx = fimage.shape[1]
+    ny = fimage.shape[0]
+    ncen = len(xinit)
+    xnew = copy.deepcopy(xinit)
+    xerr = np.zeros(ncen) + 999.
```

Comment:
We've been using a much larger value as a standard `maskval` throughout the code, so maybe best to update the numbers here too, and define `maskval=999999.9` in the function definition?

Reply:
modified.  although I don't think we are using this code currently..